### PR TITLE
Providing transformer support for sflow feature

### PR DESCRIPTION
**Pull request by Gokulnath-Raja**
**What are the changes**

Transformer support has been added for the sflow app.

**Files Changed:**

_openconfig-sampling-sflow.yang:_ which have the parameters that can be configured via gnmi and Rest and its deviations parameters has been added in the openconfig-sampling-sflow-deviation.yang

_openconfig-sampling-sflow-annot.yang:_ which consists of the transformer specific functions which needs to be called based on the rest/gnmi call.

_xfmr_sflow.go:_ go file which consists of the function definitions which needs to be called based on the subscription/get/set request and for the validations.

**Unit test file changes for Rest:**

_sflow_sonic_test.go_   - To test for the parameters in the sonic-sflow.yang

_sflow_openconfig_test.go_ – To test for the parameters in the openconfig-sampling-sflow.yang.

**Unit test file changes for Gnmi:**

translib/test/sflow

**Why the changes are required**

Currently sflow feature can be configured via click interface alone.

Transformer support has been added to support for the configuration via Rest and Gnmi

**How did I verify the changes**

Verified by adding sflow unit test files for restui and gnmi

Attached are the UT results and log. 

[gnmi_ut_for_openconfig_sflow.txt](https://github.com/sonic-net/sonic-mgmt-common/files/11591328/gnmi_ut_for_openconfig_sflow.txt)
[gnmi_ut_for_sonic_sflow.txt](https://github.com/sonic-net/sonic-mgmt-common/files/11591329/gnmi_ut_for_sonic_sflow.txt)
[rest_ut_for_sflow.txt](https://github.com/sonic-net/sonic-mgmt-common/files/11591610/rest_ut_for_sflow.txt)
 
Attached are the subscription UT  log

[target_defined_subscription.txt](https://github.com/sonic-net/sonic-mgmt-common/files/11596138/target_defined_subscription.txt)
[On_change_subscription.txt](https://github.com/sonic-net/sonic-mgmt-common/files/11596139/On_change_subscription.txt)
[Sample_subscription.txt](https://github.com/sonic-net/sonic-mgmt-common/files/11596191/Sample_subscription.txt)

**Support added:**

1. Sflow agent, collector and interface parameters can be modified via rest and gnmi.

2. Collector Name cannot be configured from rest_ui, it will be auto-generated with pattern using Collector-IP,Collector-Port & Collector-VRF.  For e.g. 10.1.1.1_6343_default 

3. Port , Portchannel , Mgmt port , Vlan can be configured as an Sflow Agent Interface. So, while configuring for the sflow agent prior configurations of the interface is mandatory.   Vlan leafref is not available now. 

4. Alias support for the interface names are not available to configure for the sflow agent and for the sflow interface. 

5. We cannot configure more than 2 collectors.

6. We need to enable the sflow globally , which interns enable the sflow in all available interfaces and then we can disable the sflow interface individually if required. We cannot disable the sflow globally and enable the interfaces locally.

### CODE REVIEW

**Comment by Gokulnath-Raja** [comment]
Following are the dependent pull request,

https://github.com/sonic-net/sonic-buildimage/pull/15226
https://github.com/sonic-net/sonic-mgmt-common/pull/89
https://github.com/sonic-net/sonic-mgmt-framework/pull/115

https://github.com/sonic-net/sonic-mgmt-common/pull/87

File: translib/test/sflow/openconfig_sflow/05_add_col.json
Comment (sachinholla) [inline translib/test/sflow/openconfig_sflow/05_add_col.json:5]: Is this a reference to openconfig's network-instance yang? It is not supported in the community master yet

File: translib/test/sflow/README.md
Comment (sachinholla) [inline translib/test/sflow/README.md:3]: Can we avoid gnmi specific test steps in sonic-mgmt-common repo?
I think these are manual test cases that require manual verification. There are gotest cases in sonic-gnmi repo, which bring up a test server and run get/set tests by reading the input from files. It is better to move these stuffs there. If you cannot enhance the sonic-gnmi test suite now, please attach these to the PR for future reference instead of committing to sonic-mgmt-common

File: translib/transformer/xfmr_sflow.go
Comment (sachinholla) [inline translib/transformer/xfmr_sflow.go:116]: Please change all these debug logs into verbose logs -- `log.V(3).Info(...)`

All logs gets redirected to main syslog in the sonic community images. These info logs will flood the syslog with useless data.

**Review by sachinholla — CHANGES_REQUESTED** [review]
Please mention that this PR depends on #87 and #89

**Comment by Gokulnath-Raja** [comment]
> Please mention that this PR depends on #87 and #89

Already updated in above comment

**Comment by Gokulnath-Raja** [comment]
> Following are the dependent pull request,

already updated in above comment (Following are the dependent pull request,)

File: models/yang/sonic/import.mk
Comment (Gokulnath-Raja) [reply models/yang/sonic/import.mk:5]: ok done... will push in next commeit

File: models/yang/annotations/openconfig-sampling-sflow-annot.yang
Comment (Gokulnath-Raja) [reply models/yang/annotations/openconfig-sampling-sflow-annot.yang:14]: ok done... will push in next commit

File: translib/Makefile
Comment (mohanapriya-meganathan) [reply translib/Makefile:58]: @amrutasali I have added this based on our discussion. Could you please provide your inputs.

Comment (amrutasali) [reply translib/Makefile:58]: transformer.test  is just  for transformer infrastructure test only. We didn't want to mix app specific code in transformer test to avoid app specific failures to be attributed to infra.

File: translib/test/sflow/README.md
Comment (Gokulnath-Raja) [reply translib/test/sflow/README.md]: We have removed the manual testing changes

File: translib/Makefile
Comment (Gokulnath-Raja) [reply translib/Makefile]: We have added the test app to pipelines in azure-pipelines.yml and removed the test app build tags

Comment (sachinholla) [reply translib/Makefile]: Files without build tags will be part of both test binaries. You cannot achieve the infra & app separation as @amrutasali indicated. Test cases will be run twice as part of both binaries (eg. search "=== RUN   Test_node_sonic_sflow_interface/Enable_sFlow" in the [PR sanity](https://dev.azure.com/mssonic/build/_build/results?buildId=285126&view=logs&j=83516c17-6666-5250-abde-63983ce72a49&t=c10d5f44-55ce-55d7-7975-407ed75d9a96&l=7519) logs). The 2nd test binary is redundant.

We have got two options:

1) Do not use the 2nd build tag for apps and keep only one test binary. The xfmrtest tag will be used to only to enable extra transformer apps when test yang models are included.

2) Use 2 build tags and 2 test binaries (the original approach used in this PR). Only problem here is that it assumes developers will add correct build tag in every *_test.go files. I feel it will be difficult to enforce this in the community. If someone does not add the tag, the test cases will be part of both tests and we will lose infra & app separation. All PR reviewers and approvers need to watch it.

File: models/yang/extensions/openconfig-sampling-sflow-deviation.yang
Comment (Gokulnath-Raja) [reply models/yang/extensions/openconfig-sampling-sflow-deviation.yang:1]: ok it will be taken care in next commit, just verified deviation is copied to target directory
root@sonic:/usr/models/yang# ls *sflow*
openconfig-sampling-sflow-annot.yang  openconfig-sampling-sflow.yang  sonic-sflow.yang
root@sonic:/usr/models/yang# exit

after change

root@sonic:/usr/models/yang# ls *sflow*
openconfig-sampling-sflow-annot.yang  openconfig-sampling-sflow-deviation.yang  openconfig-sampling-sflow.yang  sonic-sflow.yang
root@sonic:/usr/models/yang#

Comment (Gokulnath-Raja) [reply models/yang/extensions/openconfig-sampling-sflow-deviation.yang:1]: This will create conflict as it depends on https://github.com/sonic-net/sonic-mgmt-common/pull/89/files#diff-d832e5808954c6cf44be03eaa0e64428aecb80ad00129008a561c9c995e4e0fd merge first.

Comment (sachinholla) [reply models/yang/extensions/openconfig-sampling-sflow-deviation.yang:1]: That should not be a problem. For this PR tests to succeed, both #89 and #87 should be committed first and then merged to this branch. Conflict can be resolved at that time.

File: translib/Makefile
Comment (kwangsuk) [reply translib/Makefile]: Another option is to use the negation of "xfmrtest" tag to separate tests.
i.e. in the sflow_*_test.go, add the build tag
`// +build !xfmrtest`

In case a community user wants to implement new test apps, one is likely to check the build tag from existing test code as a reference.

Comment (mohanapriya-meganathan) [reply translib/Makefile:58]: As we discussed, I have proceeded with the second option of original approach used in the PR. ie using testapp build tag for the application test and added in the PR sanity pipelines.

**Review by sachinholla — APPROVED** [review]
Test binary related changes look fine. Did not review sflow app related changes.

**Review by kwangsuk — DISMISSED** [review]

**Comment by sachinholla** [comment]
hi @kwangsuk, please do not merge this PR before https://github.com/sonic-net/sonic-buildimage/pull/15226 is merged. Full image builds may fail without those buildimage makefile changes. It includes an enhancement to the main docker image makefile slave.mk, hence waiting for MSFT's approval

**Comment by sachinholla** [comment]
Hi @Gokulnath-Raja , some tests are failing -- https://dev.azure.com/mssonic/be1b070f-be15-4154-aade-b1d3bfb17054/_apis/build/builds/334869/logs/12

```
2023-08-10T05:14:43.8672835Z --- FAIL: Test_node_openconfig_sflow_interface (20.01s)
2023-08-10T05:14:43.8673189Z     --- PASS: Test_node_openconfig_sflow_interface/Test_set_on_sflow_interface (0.00s)
2023-08-10T05:14:43.8673580Z     --- PASS: Test_node_openconfig_sflow_interface/Verify_set_on_sflow_interface (0.00s)
2023-08-10T05:14:43.8673963Z     --- FAIL: Test_node_openconfig_sflow_interface/Test_delete_on_sflow_interface (0.00s)
2023-08-10T05:14:43.8674363Z     --- FAIL: Test_node_openconfig_sflow_interface/Verify_delete_on_sflow_interface (0.00s)
2023-08-10T05:14:43.8674750Z     --- FAIL: Test_node_openconfig_sflow_interface/Test_get_on_sflow_interface (0.00s)
2023-08-10T05:14:43.8675160Z     --- FAIL: Test_node_openconfig_sflow_interface/Update_admin-state_for_sflow_interface (0.00s)
2023-08-10T05:14:43.8675561Z     --- FAIL: Test_node_openconfig_sflow_interface/Verify_admin-state_for_sflow_interface (0.00s)
2023-08-10T05:14:43.8675981Z     --- FAIL: Test_node_openconfig_sflow_interface/Update_sampling-rate_on_sflow_interface (0.00s)
2023-08-10T05:14:43.8676384Z     --- FAIL: Test_node_openconfig_sflow_interface/Verify_sampling-rate_on_sflow_interface (0.00s)
2023-08-10T05:14:43.8676788Z     --- PASS: Test_node_openconfig_sflow_interface/Test_delete_on_sflow_interfaces (0.00s)
```

**Comment by mohanapriya-meganathan** [comment]
> Hi @Gokulnath-Raja , some tests are failing -- https://dev.azure.com/mssonic/be1b070f-be15-4154-aade-b1d3bfb17054/_apis/build/builds/334869/logs/12
> 
> ```
> 2023-08-10T05:14:43.8672835Z --- FAIL: Test_node_openconfig_sflow_interface (20.01s)
> 2023-08-10T05:14:43.8673189Z     --- PASS: Test_node_openconfig_sflow_interface/Test_set_on_sflow_interface (0.00s)
> 2023-08-10T05:14:43.8673580Z     --- PASS: Test_node_openconfig_sflow_interface/Verify_set_on_sflow_interface (0.00s)
> 2023-08-10T05:14:43.8673963Z     --- FAIL: Test_node_openconfig_sflow_interface/Test_delete_on_sflow_interface (0.00s)
> 2023-08-10T05:14:43.8674363Z     --- FAIL: Test_node_openconfig_sflow_interface/Verify_delete_on_sflow_interface (0.00s)
> 2023-08-10T05:14:43.8674750Z     --- FAIL: Test_node_openconfig_sflow_interface/Test_get_on_sflow_interface (0.00s)
> 2023-08-10T05:14:43.8675160Z     --- FAIL: Test_node_openconfig_sflow_interface/Update_admin-state_for_sflow_interface (0.00s)
> 2023-08-10T05:14:43.8675561Z     --- FAIL: Test_node_openconfig_sflow_interface/Verify_admin-state_for_sflow_interface (0.00s)
> 2023-08-10T05:14:43.8675981Z     --- FAIL: Test_node_openconfig_sflow_interface/Update_sampling-rate_on_sflow_interface (0.00s)
> 2023-08-10T05:14:43.8676384Z     --- FAIL: Test_node_openconfig_sflow_interface/Verify_sampling-rate_on_sflow_interface (0.00s)
> 2023-08-10T05:14:43.8676788Z     --- PASS: Test_node_openconfig_sflow_interface/Test_delete_on_sflow_interfaces (0.00s)
> ```

@sachinholla ,yes, we are looking into it.

**Comment by mohanapriya-meganathan** [comment]
@sachinholla Sanity test fails as it depends on a SFLOW_SESSION_TABLE which will be created by sflow docker container. As the azure pipelines, runs the sanity test in an isolated sonic-mgmt-framework environment it fails. So, updated the changes in the unit test framework to support the interdependency table.

Updated log:
[testapp.log](https://github.com/sonic-net/sonic-mgmt-common/files/12331827/transformer.log)

**Comment by Gokulnath-Raja** [comment]
@sachinholla please help in restarting the /AzurePipelines run Azure.sonic-common (pipeline failed due to space issue)

**Comment by sachinholla** [comment]
hi @Gokulnath-Raja, I don't know the procedure to manually trigger the PR build. I used to amend the last commit without making any changes and force push :)
```
git commit --amend --no-edit
git push -f
```

**Comment by sachinholla** [comment]
Hi Gokul, please merge latest from master branch to solve "no space left on device" error

**Comment by Gokulnath-Raja** [comment]
> Hi Gokul, please merge latest from master branch to solve "no space left on device" error

@sachinholla done and pushed the code, kindly merge...

**Comment by sachinholla** [comment]
I don't have merge permissions.. please follow up with Kwan

**Review by kwangsuk — APPROVED** [review]

**Pull request merged**
